zkevm-circuits/src/evm_circuit/execution/extcodehash.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        param::N_BYTES_ACCOUNT_ADDRESS,
+        step::ExecutionState,
+        table::{AccountFieldTag, CallContextFieldTag},
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            from_bytes, Cell, RandomLinearCombination, Word,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use eth_types::{evm_types::GasCost, Field, ToAddress, ToLittleEndian, ToScalar, U256};
+use halo2_proofs::{circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct ExtcodehashGadget<F> {
+    same_context: SameContextGadget<F>,
+    external_address: RandomLinearCombination<F, N_BYTES_ACCOUNT_ADDRESS>,
+    tx_id: Cell<F>,
+    previously_accessed: Cell<F>,
+    external_code_hash: Cell<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ExtcodehashGadget<F> {
+    const NAME: &'static str = "EXTCODEHASH";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::EXTCODEHASH;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let external_address = cb.query_rlc();
+        cb.stack_pop(external_address.expr());
+
+        let tx_id = cb.call_context(None, CallContextFieldTag::TxId);
+
+        let previously_accessed = cb.query_bool();
+        cb.account_access_list_write(
+            tx_id.expr(),
+            from_bytes::expr(&external_address.cells),
+            1.expr(),
+            previously_accessed.expr(),
+        );
+
+        let external_code_hash = cb.query_cell();
```

Comment:
Related to my previous comment about non-existing accounts:
For a non-existing account, the CodeHash is `keccack([])`, which is not 0.  But `EXTCODEHASH` pushes a 0 for a non-existing account.  So looking up the CodeHash is not enough to prove that the pushed value is correct.

I think this implementation is missing the logic to detect whether an account exists or not.

Reply:
yes, you're right. I'm going to add test cases for this as well.